glean-core/src/histogram/exponential.rs

```diff
@@ -51,142 +46,68 @@ fn exponential_range(min: u64, max: u64, bucket_count: usize) -> Vec<u64> {
-            count: 0,
-            sum: 0,
-            typ: Type::Exponential,
-        }
-    }
-
-    /// Get the number of buckets in this histogram.
-    pub fn bucket_count(&self) -> usize {
-        self.values.len()
-    }
-
-    /// Add a single value to this histogram.
-    pub fn accumulate(&mut self, sample: u64) {
-        *self.get_bucket_for_sample(sample) += 1;
-        self.sum = self.sum.saturating_add(sample);
-        self.count += 1;
-    }
-
-    /// Get the total sum of values recorded in this histogram.
-    pub fn sum(&self) -> u64 {
-        self.sum
-    }
-
-    /// Get the total count of values recorded in this histogram.
-    pub fn count(&self) -> u64 {
-        self.count
-    }
-
+impl Bucketing for PrecomputedExponential {
     /// Get the bucket for the sample.
     ///
     /// This uses a binary search to locate the index `i` of the bucket such that:
     /// bucket[i] <= sample < bucket[i+1]
-    fn get_bucket_for_sample(&mut self, sample: u64) -> &mut u64 {
+    fn sample_to_bucket_minimum(&self, sample: u64) -> u64 {
         let limit = match self.bucket_ranges.binary_search(&sample) {
             // Found an exact match to fit it in
             Ok(i) => i,
             // Sorted it fits after the bucket's limit, therefore it fits into the previous bucket
             Err(i) => i - 1,
         };
 
-        &mut self.values[limit]
-    }
-
-    /// Get the list of ranges.
-    pub fn bucket_ranges(&self) -> &[u64] {
-        &self.bucket_ranges
-    }
-
-    /// Get the filled values.
-    pub fn values(&self) -> &[u64] {
-        &self.values
+        self.bucket_ranges[limit]
     }
+}
 
-    /// Get the minimal recordable value.
-    pub fn min(&self) -> u64 {
-        self.min
-    }
+impl Histogram<PrecomputedExponential> {
+    /// Create a histogram with `count` exponential buckets in the range `min` to `max`.
+    pub fn exponential(
+        min: u64,
+        max: u64,
+        bucket_count: usize,
+    ) -> Histogram<PrecomputedExponential> {
+        let bucket_ranges = exponential_range(min, max, bucket_count);
 
-    /// Get the maximal recordable value.
-    ///
-    /// Samples bigger than  the `max` will be recorded into the overflow bucket.
-    pub fn max(&self) -> u64 {
-        self.max
+        Histogram {
+            values: HashMap::new(),
+            count: 0,
+            sum: 0,
+            bucketing: PrecomputedExponential { bucket_ranges },
+        }
     }
 
-    /// Get this histogram's type.
-    pub fn typ(&self) -> Type {
-        self.typ
-    }
+    /// Get a snapshot of _all_ values.
+    pub fn snapshot_values(&self) -> HashMap<u64, u64> {
+        let mut res = self.values.clone();
 
-    /// Check if this histogram recorded any values.
-    pub fn is_empty(&self) -> bool {
-        self.count() == 0
+        for &min_bucket in &self.bucketing.bucket_ranges {
+            // Fill in missing entries.
+            let _ = res.entry(min_bucket).or_insert(0);
```

Comment:
There's no need to fill in all values -- just from 0 to (max_non_zero_bucket + 1).  I think this is a useful optimization to reduce ping sizes.  See [here](https://github.com/mozilla/glean/blob/master/docs/dev/core/internal/payload.md#custom-distribution)